keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,326 @@
+the `Burstable` QoS level, and **all** of the containers specify limits for each resource, then these limits are summed, and that limit is
+assigned to the pod level cgroup.
+
+The implications of the current `Burstable` definition is that if any containers belonging to the Pod don't define a limit, then those containers
+are effectively not limited by the Linux kernel. Conversely, if all of the containers provided a limit, then even though the pod level cgroup is 
+configured with the sum of those limits, there is no significance to this since no container can ever use more of the resource than what is defined
+in the container level cgroup.
+
+The proposal in this KEP is to allow users to opt-in to a slightly modified definition of the `Burstable` QoS level. In this modified definition,
+the sum of all defined container limits for each resource are always assigned to the pod-level cgroup. This is done by adding an attribute called 
+`shareBurstableLimits` on the Pod specification level. If this attribute is set to false (the default) the feature is disabled, and if it is set to 
+true then the feature takes effect.
+
+For example, consider the following Pod:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: nginx
+  name: nginx
+spec:
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: proxy
+    name: envoy:latest
+  - name: nginx
+    image: nginx:latest
+    command: [ "/usr/bin/tail", "-f", "/dev/null"] 
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+This pod belongs to the `Burstable` QoS level because it defines multiple containers, of which only one has resource limits, and there is a 
+difference between the `Request` and the `Limit` for both these resources. Effectively, the containers which don't define any limits are
+not constrained in any way. They are provided the same level of resources as a container in a `BestEffort` pod would, except that they belong to a pod with a higher priority, so they are more proofed from causing evictions.
+
+The cgroup hierarchy for each of these resources (memory and cpu) would be this:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod (memory: unlimited, cpu: unlimited)
+      |
+      +-- container0 (pause container, memory: unlimited, cpu: unlimited)
+      |
+      +-- container1 (shell container, memory: unlimited, cpu: unlimited)
+      |
+      +-- container2 (proxy container, memory: unlimited, CPU: unlimited)
+      |
+      +-- container3 (nginx container, memory: 256M limit, CPU: 1 core)
+</pre>
+
+By setting the `ShareBurstableLimits` attribute on the Pod spec to `true`, the following cgroup hierarchy would be configured:
+
+<pre>
+  QoS CGroup (one of guaranteed, burstable, or besteffort)
+   |
+   \ pod (memory: 256M limit, CPU: 1 core)
```

Comment:
the biggest benefit for this is when its difficult to size two containers that may individually write to a memory backed volume.

Reply:
pod resource overhead feature is in flight, it would be good to combine this example with a pod that has a runtimeclass that has an overhead.  i also wonder as i type this if its better to associate an attribute on the runtime class to say if the pod should be bounded or not rather than change the pod spec itself.  have you explored this idea?

Reply:
Can you link to the pod resource overhead feature? I'm not familiar with this.

Intuitively the pod level seems to me to be a better fit since this sort of resource allocation is not dependent on any specific runtime, and has no meaning outside of the specific `Burstable` QoS level. But I'd love to read about the pod resource overhead feature, and give it some thought...

Reply:
While we are waiting for merge of the updated Beta docs, I'd suggest checking out the PR: https://github.com/kubernetes/website/pull/19059/files

Reply:
Thanks, I'll go over this.